JohnRoyCabezas/ps-university-ebulletin-board
# [015][BE] Add CRUD functions for Announcements in DashboardController

## Asana Link

https://app.asana.com/0/1203011106874563/1203034611151683

## Pre-requisite

- [x] setup Postgres database and update .env file in `api` folder
```
DB_CONNECTION=pgsql
DB_HOST=127.0.0.1
DB_PORT=5432
DB_DATABASE=UniversityEB
DB_USERNAME=postgres
DB_PASSWORD=postgres
```

## Commands

- [x] in the terminal, go to the project directory and cd into api directory - `cd api`
- [x] to create the tables and seed, run `php artisan migrate --seed` 
- [x] or to have a fresh migration run `php artisan migrate:fresh --seed`

## Testing as User
* A student/user only has the permission to view/read an announcement or all announcements
  - [x] Fetching all announcements
  * In an API platform like Postman, set the url to `localhost:8000/api/announcement`
  * send a GET request
   
  - [x] Fetching a specific announcement
  * make sure the $id of the announcement you are looking for is available in the announcements table
  * set the url to `localhost:8000/api/announcement/{id}`
  * send a GET request

### Expected Output
 
* fetching all announcements - the API should return the list of announcements
* fetching a specific announcement - the API should return the details of a specific announcement

### Screenshots

![image](https://user-images.githubusercontent.com/85822633/192726149-15b3006d-120c-4ea8-b66b-bca88c9c4b7e.png)
![image](https://user-images.githubusercontent.com/85822633/192726317-b763ebc8-dfb6-4cba-a704-4f2881b14203.png)

## Testing as Admin
* An Admin has CRUD permissions for announcements
  - [x] Fetching all announcements
  * In an API platform like Postman, set the url to `localhost:8000/api/announcement`
  * send a GET request
   
  - [x] Fetching a specific announcement
  * make sure the $id of the announcement you are looking for is available in the announcements table
  * set the url to `localhost:8000/api/announcement/{id}`
  * send a GET request

  - [x] Creating an announcement
  * set the url to `localhost:8000/api/announcement`
  * input the needed params: `announcementable_id : 1`, `announcementable_type: App/Model/College`, `announcement: This is an announcement for the Engineering College!`
  * send a POST request

  - [x] Updating an announcement
  * set the url to `localhost:8000/api/announcement/{id}`
  * make sure the $id of the announcement you are trying to update is available in the announcements table
  * input the needed params: `announcement_update: This is an update message for the announcement!`
  * send a PUT request

  - [x] Deleting an announcement
  * set the url to `localhost:8000/api/announcement/{id}`
  * make sure the $id of the announcement you are trying to delete is available in the announcements table
  * send a DELETE request

### Expected Output
 
* fetching all announcements - the API should return the list of announcements
* fetching a specific announcement - the API should return the details of a specific announcement
* creating an announcement - the API should create a new item on the announcements table
* updating an announcement - the API should update an announcement in the announcements table
* deleting an announcement - the API should remove a given announcement from the announcements table

### Screenshots
![image](https://user-images.githubusercontent.com/85822633/192728938-a488d309-00ae-4778-a190-7c1f72f7434e.png)
![image](https://user-images.githubusercontent.com/85822633/192729474-a285088a-b8d5-4b59-ab60-d8ce0f6396b0.png)
![image](https://user-images.githubusercontent.com/85822633/192729094-6e941688-e8fd-4afd-bcb5-f24dbb64e16f.png)
![image](https://user-images.githubusercontent.com/85822633/192729731-33cc6867-d4d9-4a55-8a16-a0b926bd5685.png)
![image](https://user-images.githubusercontent.com/85822633/192729193-809cb6db-fa95-481e-b6bc-076abebb7ad3.png)
